go/host/db/blocks.go

```diff
@@ -2,61 +2,148 @@ package db
 
 import (
 	"bytes"
+	"errors"
 	"fmt"
-
-	"github.com/obscuronet/go-obscuro/go/common/errutil"
+	"math/big"
 
 	gethcommon "github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/ethdb"
 	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/errutil"
 )
 
-// DB methods relating to blocks.
+// DB methods relating to block headers.
 
-// GetBlockHeader returns the block header given the hash.
-func (db *DB) GetBlockHeader(hash gethcommon.Hash) (*types.Header, error) {
-	return db.readBlockHeader(db.kvStore, hash)
+// GetBlockByHash returns the block header given the hash.
+func (db *DB) GetBlockByHash(hash gethcommon.Hash) (*types.Header, error) {
+	return db.readBlock(db.kvStore, blockHashKey(hash))
 }
 
-// AddBlockHeader adds a types.Header to the known headers
-func (db *DB) AddBlockHeader(header *types.Header) error {
+// GetBlockByHeight returns the block header given the height
+func (db *DB) GetBlockByHeight(height *big.Int) (*types.Header, error) {
+	return db.readBlock(db.kvStore, blockNumberKey(height))
+}
+
+// AddBlock adds a types.Header to the known headers
+func (db *DB) AddBlock(header *types.Header) error {
 	b := db.kvStore.NewBatch()
-	err := db.writeBlockHeader(header)
+	err := db.writeBlockByHash(header)
+	if err != nil {
+		return fmt.Errorf("could not write block header. Cause: %w", err)
+	}
+
+	err = db.writeBlockByHeight(header)
 	if err != nil {
 		return fmt.Errorf("could not write block header. Cause: %w", err)
 	}
 
+	// Update the tip if the new height is greater than the existing one.
+	tipBlockHeader, err := db.GetBlockAtTip()
+	if err != nil && !errors.Is(err, errutil.ErrNotFound) {
+		return fmt.Errorf("could not retrieve block header at tip. Cause: %w", err)
+	}
+	if tipBlockHeader == nil || tipBlockHeader.Number.Cmp(header.Number) == -1 {
+		err = db.writeBlockAtTip(b, header.Hash())
+		if err != nil {
+			return fmt.Errorf("could not write new block hash at tip. Cause: %w", err)
+		}
+	}
+
 	if err = b.Write(); err != nil {
 		return fmt.Errorf("could not write batch to DB. Cause: %w", err)
 	}
 
 	return nil
 }
 
-// headerKey = blockHeaderPrefix  + hash
-func blockHeaderKey(hash gethcommon.Hash) []byte {
-	return append(blockHeaderPrefix, hash.Bytes()...)
+// GetBlockListing returns a list of blocks given the pagination
+func (db *DB) GetBlockListing(pagination *common.QueryPagination) (*common.BlockListingResponse, error) {
```

Comment:
Finding the use of `Listing` a bit confusing, sometimes it seems to refer to a list and sometimes it seems to refer to an entry in that list. Might be good to try and make the naming convention a bit more consistent if poss.

Reply:
It's meant to always be a list, but where are you seeing being referred as an entry ?

Reply:
`BlockListingResponse` contains a list of Listings

Reply:
Fixed, thanks for the call 👏